src/System.Diagnostics.Tracing/src/System/Diagnostics/Tracing/EventCounter.cs

```diff
@@ -264,7 +286,8 @@ private void OnEventSourceCommand(object sender, EventCommandEventArgs e)
                 float value;
                 if (e.Arguments.TryGetValue("EventCounterIntervalSec", out valueStr) && float.TryParse(valueStr, out value))
                 {
-                    EnableTimer(value);
+                    lock (this)      // Lock the EventCounterGroup
+                        EnableTimer(value);
```

Comment:
It is true that EnableTimer can call the Timer callback, and that in turn logs EventSource events and that there may be user code that has subscribed to the EventSource via an EventListener called when that EventSource.Write call is made.  

 But this is just no different than any other call to EventSource.Write, and we don't have a prohibition against calling EventSource.Write while holding a lock (that would be very constraining).    Instead we place the burden on the EventListener to avoid locking issues.  

In this case it should be easy for an EventListener to avoid a problem.   Basically the chain of events to form a recursion is that some EventListener would have to call 'Enable' on an EventSource which in turns on an EventSources' eventCounters.  There would be some EventListener subscribing to that EventSource that would get a 'OnWrite' callback.  This OnWrite callback would have to take a lock.  Then there would be code (on another thread) that takes this OnWrite lock, and calls EventListener.Enable, which would cause the deadlock.  Thus there would have to be code 'beneath' 'OnWrite' (that is it takes the OnWrite lock), that calls EventListener.Enable (that is it would want to poll eventCounters or change the sampling of EventCounters).  While this COULD certainly happen, it a pretty unlikely scenario, and it would be relatively easy for the OnWrite code to drop any it held before calling EventListener.Enable.   

So I think you can make the case that other code should worry about the lock loops.  However it certainly would be better if you did not even need to think about this.  To do this we can either

1.  Drop locks before calling EventSource.Write().  We can do this but we would probably need to clone the list of EventCounters and the code does get more complex and subtle. 

2.  Don't call OnTimer synchonrous when the EventCounters are enabled.   This is trivial, but it does make a polling scenario not nearly as nice (you can force a poll, but you have to wait in indeterminate time to actually get it).    I think supporting the synchronous case (after calling Enable and it returns you KNOW that you should have received a EventCounter payload).  

I am willing to fix this any way you think is best, but my inclination right now is to do nothing (effectively make it the EventListener's issue), because I think that the scenario that could cause deadlock is not likely to ever be interesting, and if it is, the EventListener is in the best position to fix it.   

Let me know how you want to handle this.